### PR TITLE
Add tensorflow implementation of hexagonal convolutions

### PR DESCRIPTION
**Pull request by lgeiger**
@ehoogeboom, @jornpeters, @tscohen Thanks for writing the paper and making this code public!

We're really interested in applying hexagonal group equivariant convolutions for event reconstruction in astroparticle physics since those problems are often invariant under rotation and signals are usually measured by detectors  arranged in a hexagonal grid.

This PR aims at implementing support for [Tensorflow](tensorflow.org) using the new [`tf.layers` API](https://www.tensorflow.org/api_docs/python/tf/layers/Layer).

This isn't ready be merged yet but I want to share the progress so far in order to get early feedback and prevent duplicated work.
The core of this PR is adapted from [GrouPy](https://github.com/tscohen/GrouPy/tree/master/groupy/gconv/tensorflow_gconv).
~~This PR is based on #1 so the first two commits aren't relevant for review here.~~ Rebased on top of master.

## Usage Example
```python
import numpy as np
import tensorflow as tf
from groupy.gconv.gconv_tensorflow.layers import P6ConvZ2Axial, P6ConvP6Axial

x = tf.placeholder(tf.float32, [None, 9, 9, 3])
conv1 = P6ConvZ2Axial(filters=64, kernel_size=3, use_bias=False, padding='same')(x)
conv2 = P6ConvP6Axial(filters=64, kernel_size=3, use_bias=False, padding='same')(conv1)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    output1 = sess.run(conv1, feed_dict={x: np.random.randn(10, 9, 9, 3)})
    output = sess.run(conv2, feed_dict={x: np.random.randn(10, 9, 9, 3)})

print(output1.shape)  # (10, 9, 9, 384)
print(output.shape)  # (10, 9, 9, 384)
```

## TODO
- [x] Add unit tests for index transformation
- [x] Add unit tests for p4/p4m layers and check equivariance
- [x] Add unit tests for p6/p6m layers and check equivariance
- [x] Add support for bias
- [x] Add Keras layers
- [x] Add documentation

## Ideas for future PRs
- Add functional interface as [present in `tf.layers`](https://www.tensorflow.org/api_docs/python/tf/layers/conv2d)
- Benchmark performance and check if there's room for improvement
- Add support and unit tests for `NCHW` format (I probably won't work on this 😄)

### CODE REVIEW

**Comment by lgeiger** [comment]
This PR is now ready for people to check out. 🎉 

There is still on minor problem, I'm unsure about:
For some reason when using a non zero bias together with hexagonal convolutions the equivariance tests fail. I'm not sure if this is expected or a bug somewhere in the code.
Tomorrow I'll try to get the `gconv_chainer` unittests working to see if the issue is also present there. I haven't worked with `chainer` but to me it looks like the unittests use only a zero bias. I'd really appreciate some feedback on this.

Here is a small script that trains a very simple model on the CIFAR-10 dataset using the `P4Conv` [Keras](keras.io) layers introduced in this PR: https://gist.github.com/lgeiger/f4cba03c87f8f036d6a1c11f2ae962ba

**Comment by tscohen** [comment]
Hi Lukas, thanks a lot for your contribution! Very exciting to see our work being put to such a great use.

I had a brief look at your code, and it looks great, so we can probably merge soon. Regarding the bias issue: as I think you know, adding a bias is only equivariant if you add the same value to each of the 4/6/8/12 orientation-channels of a single G-channel. It does seem like that is what you are doing, so there is probably some subtle bug. Is this still an issue, or was it resolved in your last few commits?

**Comment by lgeiger** [comment]
Hi Taco thanks for the response.

> I had a brief look at your code, and it looks great, so we can probably merge soon.

👍Sorry for the amount of added lines, most of them are just docstrings and there's no nice way of handling them in subclasses. 

> Regarding the bias issue: Is this still an issue, or was it resolved in your last few commits?

This is still a issue. I just checked the `chainer` implementation and I can reproduce the same issue by adding `initial_bias=np.random.rand(out_channels)` to the unit tests (works for P4/p4m but not for p6/p6m).
So this is likely not a bug in the tensorflow implementation. I suspect there's something is wrong with the way we check equivariance or a very subtle bug somewhere else.

The last commits are aimed to bring both implementations closer together and make the code more readable. In theory this should also bring a performance boost but I wasn't able to benchmark this yet.

Note that 28cb3272cb1d10ef1e767ac6817391acbefa46e7 requires at least Tensorflow 1.4 because of the usage of `tf.gather_nd` but I think the advantages of this approach outway the incompatibility with legacy versions of Tensorflow.

**Comment by ehoogeboom** [comment]
Hi Lukas,

Thank you for your interest in our paper and code. Your contribution looks great and we would like to merge soon. 

Regarding the bias issue:
- Throughout our code we assume a _rectangular_ image shape, inside a hexagonal lattice (e.g. CIFAR-10 or AID images). However, only _hexagonal_ shapes inside hexagonal lattices can truly be equivariant. 
- You can see this, by imagining what happens when you rotate an image from Fig. 6b of our paper. Unless you rotate it a multiple of 180 degrees, some non-zero pixels will fall outside the boundary. 
- For this reason exactly, there is a function `hex_padding` in `test_gconv.py`. We create input images with large zero borders to test equivariance. At the boundaries the network is not entirely equivariant. 
- Consequently, when a bias is added, values at the boundary are nonzero, and thus the network will not be completely equivariant.

Alternatively, we could have defined larger hexagonal shapes in hexagonal lattices, that would fit the rectangular shape, but this would have incurred a larger performance penalty as the feature map contains more zero-padding. Nonetheless, this would have made the network exactly equivariant for the groups p6 and p6m. 

In summary, it is expected that unit tests fail if a non-zero bias term is added.

**Comment by lgeiger** [comment]
@ehoogeboom Thanks for the detailed explanation. I was naively expecting the output mask would take care of this.

I removed the related comments from the unit tests.

**Comment by ehoogeboom** [comment]
@lgeiger Thanks for your efforts. Is it perhaps possible to reproduce our experiments on AID, to validate that everything is working correctly?

**Comment by dqii** [comment]
Is G-Pooling included in this implementation?

**Comment by lgeiger** [comment]
Thanks for reviewing.
I added [unittests and CI](https://travis-ci.org/lgeiger/hexaconv/builds/349410229) to ensure that there are no major regressions. But I agree a full reproduction of the experiments would be great.
Unfortunately I probably won't have time to reproduce the experiments from your paper with this branch. It would be great if someone else could take this for a spin.
